kexoqs.c

```diff
@@ -480,51 +324,50 @@ int
 oqs_server_gen_msg_and_ss(OQS_KEX_CTX *oqs_kex_ctx,
 	u_char **oqs_shared_secret, size_t *oqs_shared_secret_len) {
 
-	OQS_KEX *oqs_kex = NULL;
+	OQS_KEM *oqs_kem = NULL;
 	uint8_t *tmp_oqs_shared_secret = NULL;
-	size_t tmp_oqs_shared_secret_len = 0;
 	int r = 0;
 
-	if (OQS_NEED_SEED == oqs_kex_ctx->oqs_need_seed) {
-		if (oqs_kex_ctx->oqs_seed == NULL ||
-			oqs_kex_ctx->oqs_seed_len != oqs_kex_ctx->oqs_need_seed_len) {
-			fprintf(stderr, "seed failed\n");
-			r = SSH_ERR_INTERNAL_ERROR;
-			goto out;
-		}
+	if ((oqs_kem = OQS_KEM_new(oqs_kex_ctx->oqs_method)) == NULL) {
+		r = SSH_ERR_INTERNAL_ERROR;
+		goto out;
 	}
-	else {
-		oqs_kex_ctx->oqs_seed = NULL;
-		oqs_kex_ctx->oqs_seed_len = 0;
+
+	if ((oqs_kex_ctx->oqs_local_msg = malloc(oqs_kem->length_ciphertext)) == NULL) {
+		r = SSH_ERR_ALLOC_FAIL;
+		goto out;
+	}
+	if ((tmp_oqs_shared_secret = malloc(oqs_kem->length_shared_secret)) == NULL) {
+		r = SSH_ERR_ALLOC_FAIL;
+		goto out;
 	}
 
-	if ((oqs_kex = OQS_KEX_new(&oqs_rnd, oqs_kex_ctx->oqs_method,
-		oqs_kex_ctx->oqs_seed, oqs_kex_ctx->oqs_seed_len,
-		oqs_kex_ctx->oqs_param)) == NULL) {
-		r = SSH_ERR_INTERNAL_ERROR;
+	if (oqs_kex_ctx->oqs_remote_msg_len != oqs_kem->length_public_key) {
```

Comment:
This could also be done before allocating memory for the local msg and shared secret.